src/error.rs

```diff
@@ -203,16 +199,67 @@ impl<'input, 'state> ParsedFields<'input, 'state> {
         Some(quote!(#pattern => #expr))
     }
 
-    fn render_backtrace_as_struct(&self) -> Option<TokenStream> {
+    fn render_provide_as_struct(&self) -> Option<TokenStream> {
         let backtrace = self.backtrace?;
-        let backtrace_expr = &self.data.members[backtrace];
-        Some(quote!(Some(&#backtrace_expr)))
+
+        let source_provider = self.source.map(|source| {
+            let source_expr = &self.data.members[source];
+            quote! {
+                ::std::error::Error::provide(&#source_expr, demand);
+            }
+        });
+        let backtrace_provider = self
+            .source
+            .filter(|source| *source == backtrace)
+            .is_none()
+            .then(|| {
+                let backtrace_expr = &self.data.members[backtrace];
+                quote! {
+                    demand.provide_ref::<std::backtrace::Backtrace>(&#backtrace_expr);
```

Comment:
Food for thought: One of the reasons for this provider API (afaict) is so more types can be provided that just Backtrace, without needing to add more methods to the `Error` trait. 

So probably we want to allow this too, e.g. to embed some additionaly context such as trace IDs in there for distributed tracing. This would probabyl be better as a separate PR. I think it makes sense to keep dedicated and easy to use support for the `Backtrace` type.

Another idea is to also provide the `source` through the provider API, like is done in the rust docs

```rust
impl std::error::Error for Error {
    fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
        demand
            .provide_ref::<MyBacktrace>(&self.backtrace)
            .provide_ref::<dyn std::error::Error + 'static>(&self.source);
    }
}
```
Source: https://doc.rust-lang.org/nightly/std/error/trait.Error.html

Reply:
@JelteF I've already [raised question of what should be provided][1] and this implementation mirrors [`thiserror`][2]. As `Provider` API is quite new and `source()` default impl still returns `None`, I think that we shouldn't rush with providing `dyn Error` and wait how `Provider` API will be used throughout the ecosystem. 




[1]: https://github.com/JelteF/derive_more/issues/195#issuecomment-1231616625
[2]: https://github.com/dtolnay/thiserror/pull/182